text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,659 @@
+> an [operation name](#operation-name). The destination name MAY be appended if
+> it is of low cardinality.
+
+##### Examples
+
+* `kafka publish shop.orders`
+* `rabbitmq receive print_jobs`
+* `AmazonSQS deliver`
+* `activemq settle`
+
+#### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One ore more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more message are passed to a consumer. |
+| `settle`       | One or more message are settled. |
+
+For further details about each of those operations refer, to the [section about trace structure](#trace-structure).
+
+#### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumer and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need of additional semantic hints.
+
+#### Attributes
+
+Attribute             | Type   | Requirement Level
+----------------------|--------|---------
+[`messaging.system`](#messagingsystem) | string | Yes
+[`messaging.operation`](#messagingoperation) | string | Required
+[`messaging.destination.name`](#messagingdestinationname) | string | For producer spans
+[`messaging.destination.template`](#messagingdestinationtemplate) | string | No
+[`messaging.destination.kind`](#messagingdestinationkind) | string | No
+[`messaging.destination.temporary`](#messagingdestinationtemporary) | string | No
+[`messaging.destination.anonymous`](#messagingdestinationanonymous) | string | No
+[`messaging.source.name`](#messagingsourcename) | string | For consumer spans
+[`messaging.source.template`](#messagingsourcetemplate) | string | No
+[`messaging.source.kind`](#messagingsourcekind) | string | No
+[`messaging.source.temporary`](#messagingsourcetemporary) | string | No
+[`messaging.source.anonymous`](#messagingsourceanonymous) | string | No
+[`net.app.protocol.name`](#netappprotocolname) | string | No
+[`net.app.protocol.version`](#netappprotocolversion) | string | No
+[`net.peer.ip`](#netpeerip) | string | No
+[`net.peer.name`](#netpeername) | string | No
```

Comment:
I want to suggest a new "concept" here, which is that of the "receipt handle". In AWS SQS, when you call the [ReceiveMesssage API](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html), you get, along with the message ID (which is also missing as an attribute here), a unique "Receipt handle" for each message. The receipt handle is what you need to provide to [DeleteMessage](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessage.html) or [DeleteMessageBatch](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_DeleteMessageBatch.html) to settle the message (instead of, e.g., the message ID).

The receipt handle is a rather but not terribly long base64 string. I wonder if:
* Other messaging systems have a similar mechanism (if not, all the other questions are irrelevant at this point)
* This concept should be described in the "individual message settlement" section 
* It would make sense defining an attribute for it, e.g. to correlate with logs. It is possible that this only makes sense in a future step together with intermediary instrumentation. But it could be interesting to track whether a message is re-processed without being re-delivered/received. Maybe too much of an edge case 🙂

Reply:
Tbh I have never used AWS SQS and don't know any other messaging system having this concept, can you mention another one please?

Reply:
I don't know one, that's why I'm asking.

Reply:
I read that wrong ;-) Now it let me think that something like that should be in the Service Bus as well but in general in AMQP 1.0.
Let's take a look at the REST interface for Service Bus via HTTP. Other than a receive and delete message, there is a two-steps way: peek lock and delete. During the [Peek Lock](https://docs.microsoft.com/en-us/rest/api/servicebus/peek-lock-message-non-destructive-read) the consumer gets the message and a `sequencenumber` with it. This will be used in the next [Delete Message](https://docs.microsoft.com/en-us/rest/api/servicebus/delete-message) together with the message id to delete the message.
But even from an AMQP 1.0 point of view it makes sense because looking at the [specification](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-complete-v1.0-os.pdf), the `disposition` frame (which is the ack from the consumer) defines a `delivery-id`. This maps with the Azure Service Bus AMQP 1.0 implementation [here](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-request-response) where you can search for the sequence-number again for deleting a message. I guess that the `sequence number` maps to the `delivery-id` at AMQP 1.0 level.
Said that, I would agree that maybe a concept like a receipt handle, sequence number or delivery id could make sense to add for the "settlement" by the consumer.